src/mlpack/core/data/load_impl.hpp

```diff
@@ -164,13 +164,25 @@ bool Load(const std::string& filename,
   if (loadType != FileType::HDF5Binary)
   {
     if (loadType == FileType::CSVASCII)
-      success = loader.LoadNumericCSV(matrix, stream);
+    {
+      arma::field<std::string> headers;
```

Comment:
I've been thinking about the API to expose this to a user.  I think we could add `arma::field<std::string>& names` as an option to `data::Load()`, and document it:
 * if `names` is passed *and* the type specified is `FileType::CSVASCII` or the detected type is a CSV, then the CSV will be assumed to have a header line and column names will be stored in `names`
 * if the type is not CSV, then `names` is ignored

Since we can't specify a default value for a reference argument, I guess we'll have to have two definitions, but that's easy enough.

What do you think?  I don't think that's the only way to do it, but it's what came to my head looking at it today.  My only disappointment with the idea above is that it doesn't provide a way for the user to specify whether or not there is a header---there's not a very ergonomic way to say "the CSV has a header but I don't care what the column names area".  Maybe you have a better idea?

Reply:
Here's another idea for an API:

```
struct LoadOptions
{
  // this will need to be cleaned up for sure...
  bool fatal = false;
  bool transpose = false;
  FileType format = AutoDetect;
  arma::csv_opts armaCsvOptions;
  bool hasHeaders = false;
  arma::field<std::string> headers;
  ...
};

template<typename eT>
bool Load(const std::string& filename,
          arma::Mat<eT>& matrix,
          const bool fatal,
          const bool transpose)
{
  LoadOptions opts; // defaults
  opts.fatal = fatal;
  opts.transpose = transpose;
  return Load(filename, matrix, opts);
}

template<typename eT>
bool Load(const std::string& filename,
          arma::Mat<eT>& matrix,
          const LoadOptions& opts)
{
  ...
}
```